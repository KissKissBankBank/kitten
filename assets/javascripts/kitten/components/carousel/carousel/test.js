import React from 'react'
import renderer from 'react-test-renderer'
import {
  Carousel,
  getNumColumnsForWidth,
  getNumPagesForColumnsAndDataLength,
  checkPage,
} from './carousel'
import {
  ProjectCard,
  MIN_WIDTH as ProjectCardMinWidth,
  MARGIN_BETWEEN as ProjectCardMarginBetween,
} from '../../../components/cards/project-card'

const createMockMediaMatcher = matches => () => ({
  matches,
  addListener: () => {},
  removeListener: () => {},
})

describe('<Carousel />', () => {
  let originalMatchMedia
  const data = [{ title: 'A' }]

  beforeEach(() => {
    originalMatchMedia = window.matchMedia
  })

  afterEach(() => {
    window.matchMedia = originalMatchMedia
  })

  describe('by default on desktop', () => {
    window.matchMedia = createMockMediaMatcher(false) // desktop
    const carousel = renderer
      .create(
        <Carousel
          itemMinWidth={ProjectCardMinWidth}
          baseItemMarginBetween={ProjectCardMarginBetween}
        >
          {data.map((item, index) => (
            <ProjectCard title={item.title} key={index} />
          ))}
        </Carousel>,
      )
      .toJSON()

    it('matches with snapshot', () => {
      expect(carousel).toMatchSnapshot()
    })
  })

  describe('by default on mobile', () => {
    window.matchMedia = createMockMediaMatcher(true) // mobile
    const carousel = renderer
      .create(
        <Carousel
          itemMinWidth={ProjectCardMinWidth}
          baseItemMarginBetween={ProjectCardMarginBetween}
          pagesClassName="custom-class"
          hidePagination
        >
          {data.map((item, index) => (
            <ProjectCard title={item.title} key={index} />
          ))}
        </Carousel>,
      )
      .toJSON()

    it('matches with snapshot', () => {
      expect(carousel).toMatchSnapshot()
    })
  })

  describe('getNumColumnsForWidth', () => {
    it('5 columns', () => {
      expect(getNumColumnsForWidth(1000, 150, 50)).toBe(5)
    })

    it('1 column', () => {
      expect(getNumColumnsForWidth(300, 150, 20)).toBe(1)
    })

    it('0 column if no width', () => {
      expect(getNumColumnsForWidth(0, 100, 10)).toBe(0)
    })

    it('0 column if no itemWidth', () => {
      expect(getNumColumnsForWidth(800, 0, 0)).toBe(0)
    })

    it('NaN if not number', () => {
      expect(getNumColumnsForWidth('0', '0', '0')).toBeNaN()
    })
  })

  describe('getNumPagesForColumnsAndDataLength', () => {
    it('3 pages', () => {
      expect(getNumPagesForColumnsAndDataLength(7, 3)).toBe(3)
    })

    it('1 page', () => {
      expect(getNumPagesForColumnsAndDataLength(2, 2)).toBe(1)
    })

    it('0 page if no dataLength', () => {
      expect(getNumPagesForColumnsAndDataLength(0, 3)).toBe(0)
    })

    it('0 page if no numColumns', () => {
      expect(getNumPagesForColumnsAndDataLength(5, 0)).toBe(0)
    })

    it('NaN if not number', () => {
      expect(getNumPagesForColumnsAndDataLength('0', '0')).toBeNaN()
    })
  })

  describe('checkPage', () => {
    it('to page number 2', () => {
      expect(checkPage(4, 2)).toBe(2)
    })

    it('to page number 3', () => {
      expect(checkPage(4, 3)).toBe(3)
    })

    it('stay page number 3', () => {
      expect(checkPage(4, 4)).toBe(3)
    })

    it('to page number 1', () => {
      expect(checkPage(4, 1)).toBe(1)
    })

    it('to page number 0', () => {
      expect(checkPage(4, 0)).toBe(0)
    })

    it('stay page number 0 if newPage negative', () => {
      expect(checkPage(4, -1)).toBe(0)
    })

    it('stay page number 0 if no numPages', () => {
      expect(checkPage(0, 3)).toBe(0)
    })

    it('stay page 0 if not number', () => {
      expect(checkPage('0', '0')).toBe(0)
    })
  })
})
