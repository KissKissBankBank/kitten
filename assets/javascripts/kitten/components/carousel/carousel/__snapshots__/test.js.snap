// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Carousel /> by default on desktop matches with snapshot 1`] = `
<div
  className="sc-fzozJi dfRDUK k-Carousel"
>
  <div
    className="sc-AxgMl jWGKu k-Carousel__inner"
    onScroll={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="sc-AxheI fsZQch k-Carousel__inner__pageContainer"
      onClick={[Function]}
    >
      <div
        className="sc-AxhCb ffaSdK k-Carousel__page"
      >
        <div
          className="sc-AxhUy hZqsUu k-Carousel__page__item"
        >
          <div
            className="k-ProjectCard k-Card k-Card--light"
            href={null}
            title={null}
          >
            <div
              className="k-ProjectCard__grid"
            >
              <div
                className="sc-fzqNJr fSnHqq k-ProjectCard__grid--flex"
              >
                <span
                  className="k-ButtonImage k-ButtonImage--withoutPointerEvents"
                  tabIndex={null}
                >
                  <img
                    alt=""
                    className="k-ButtonImage__img"
                    src={null}
                  />
                </span>
                <p
                  className="k-Paragraph k-ProjectCard__grid--flex__item-fluid k-u-margin-left-single k-u-margin-right-single k-Paragraph--quaternary k-Paragraph--withoutMargin k-Paragraph--normalLineHeight"
                >
                  <span
                    className="k-u-weight-regular"
                  >
                    Name
                  </span>
                </p>
              </div>
              <div
                className="sc-fzqNJr jsUow k-ProjectCard__grid--flex"
              >
                <p
                  className="sc-fzoyAV hEKEUl k-ProjectCard__grid--flex__item-fluid k-Card__title"
                >
                  A
                </p>
              </div>
            </div>
          </div>
        </div>
        <div
          className="sc-AxhUy ezsiGz k-Carousel__page__item"
        />
        <div
          className="sc-AxhUy ezsiGz k-Carousel__page__item"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<Carousel /> by default on mobile matches with snapshot 1`] = `
<div
  className="sc-fzozJi dfRDUK k-Carousel"
>
  <div
    className="sc-AxgMl jWGKu k-Carousel__inner"
    onScroll={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="sc-AxheI fsZQch k-Carousel__inner__pageContainer"
      onClick={[Function]}
    >
      <div
        className="sc-AxhCb ffaSdK k-Carousel__page"
      >
        <div
          className="sc-AxhUy hZqsUu k-Carousel__page__item"
        >
          <div
            className="k-ProjectCard k-Card k-Card--light"
            href={null}
            title={null}
          >
            <div
              className="k-ProjectCard__grid"
            >
              <div
                className="sc-fzqNJr fSnHqq k-ProjectCard__grid--flex"
              >
                <span
                  className="k-ButtonImage k-ButtonImage--withoutPointerEvents"
                  tabIndex={null}
                >
                  <img
                    alt=""
                    className="k-ButtonImage__img"
                    src={null}
                  />
                </span>
                <p
                  className="k-Paragraph k-ProjectCard__grid--flex__item-fluid k-u-margin-left-single k-u-margin-right-single k-Paragraph--quaternary k-Paragraph--withoutMargin k-Paragraph--normalLineHeight"
                >
                  <span
                    className="k-u-weight-regular"
                  >
                    Name
                  </span>
                </p>
              </div>
              <div
                className="sc-fzqNJr jsUow k-ProjectCard__grid--flex"
              >
                <p
                  className="sc-fzoyAV hEKEUl k-ProjectCard__grid--flex__item-fluid k-Card__title"
                >
                  A
                </p>
              </div>
            </div>
          </div>
        </div>
        <div
          className="sc-AxhUy iLqwOC k-Carousel__page__item"
        />
        <div
          className="sc-AxhUy iLqwOC k-Carousel__page__item"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<Carousel /> empty carousel matches with snapshot 1`] = `null`;

exports[`<Carousel /> legacy carousel on desktop matches with snapshot 1`] = `
<div
  className="sc-AxmLO lbZFtJ k-Carousel k-LegacyCarousel k-LegacyCarousel--inGrid"
>
  <div
    className="k-Grid"
  >
    <div
      className="k-Grid__col--10 k-Grid__col--offset-1"
    >
      <div
        className="sc-AxgMl jWGKu k-Carousel__inner"
        onScroll={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
      >
        <div
          className="sc-AxheI fsZQch k-Carousel__inner__pageContainer"
          onClick={[Function]}
        >
          <div
            className="sc-AxhCb ffaSdK k-Carousel__page"
          >
            <div
              className="sc-AxhUy hZqsUu k-Carousel__page__item"
            >
              <div
                className="k-ProjectCard k-Card k-Card--light"
                href={null}
                title={null}
              >
                <div
                  className="k-ProjectCard__grid"
                >
                  <div
                    className="sc-fzqNJr fSnHqq k-ProjectCard__grid--flex"
                  >
                    <span
                      className="k-ButtonImage k-ButtonImage--withoutPointerEvents"
                      tabIndex={null}
                    >
                      <img
                        alt=""
                        className="k-ButtonImage__img"
                        src={null}
                      />
                    </span>
                    <p
                      className="k-Paragraph k-ProjectCard__grid--flex__item-fluid k-u-margin-left-single k-u-margin-right-single k-Paragraph--quaternary k-Paragraph--withoutMargin k-Paragraph--normalLineHeight"
                    >
                      <span
                        className="k-u-weight-regular"
                      >
                        Name
                      </span>
                    </p>
                  </div>
                  <div
                    className="sc-fzqNJr jsUow k-ProjectCard__grid--flex"
                  >
                    <p
                      className="sc-fzoyAV hEKEUl k-ProjectCard__grid--flex__item-fluid k-Card__title"
                    >
                      A
                    </p>
                  </div>
                </div>
              </div>
            </div>
            <div
              className="sc-AxhUy ezsiGz k-Carousel__page__item"
            />
            <div
              className="sc-AxhUy ezsiGz k-Carousel__page__item"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="k-Grid__col--1"
    />
  </div>
</div>
`;

exports[`<Carousel /> legacy carousel on desktop with withoutLeftOffset={true} matches with snapshot 1`] = `
<div
  className="sc-AxmLO lbZFtJ k-Carousel k-LegacyCarousel k-LegacyCarousel--inGrid"
>
  <div
    className="k-Grid"
  >
    <div
      className="k-Grid__col--11 k-Grid__col--offset-0"
    >
      <div
        className="sc-AxgMl jWGKu k-Carousel__inner"
        onScroll={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
      >
        <div
          className="sc-AxheI fsZQch k-Carousel__inner__pageContainer"
          onClick={[Function]}
        >
          <div
            className="sc-AxhCb ffaSdK k-Carousel__page"
          >
            <div
              className="sc-AxhUy hZqsUu k-Carousel__page__item"
            >
              <div
                className="k-ProjectCard k-Card k-Card--light"
                href={null}
                title={null}
              >
                <div
                  className="k-ProjectCard__grid"
                >
                  <div
                    className="sc-fzqNJr fSnHqq k-ProjectCard__grid--flex"
                  >
                    <span
                      className="k-ButtonImage k-ButtonImage--withoutPointerEvents"
                      tabIndex={null}
                    >
                      <img
                        alt=""
                        className="k-ButtonImage__img"
                        src={null}
                      />
                    </span>
                    <p
                      className="k-Paragraph k-ProjectCard__grid--flex__item-fluid k-u-margin-left-single k-u-margin-right-single k-Paragraph--quaternary k-Paragraph--withoutMargin k-Paragraph--normalLineHeight"
                    >
                      <span
                        className="k-u-weight-regular"
                      >
                        Name
                      </span>
                    </p>
                  </div>
                  <div
                    className="sc-fzqNJr jsUow k-ProjectCard__grid--flex"
                  >
                    <p
                      className="sc-fzoyAV hEKEUl k-ProjectCard__grid--flex__item-fluid k-Card__title"
                    >
                      A
                    </p>
                  </div>
                </div>
              </div>
            </div>
            <div
              className="sc-AxhUy ezsiGz k-Carousel__page__item"
            />
            <div
              className="sc-AxhUy ezsiGz k-Carousel__page__item"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="k-Grid__col--1"
    />
  </div>
</div>
`;

exports[`<Carousel /> legacy carousel on mobile matches with snapshot 1`] = `
<div
  className="sc-AxmLO lbZFtJ k-Carousel k-LegacyCarousel k-LegacyCarousel--noGrid"
>
  <div
    className="sc-AxgMl jWGKu k-Carousel__inner"
    onScroll={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="sc-AxheI fsZQch k-Carousel__inner__pageContainer"
      onClick={[Function]}
    >
      <div
        className="sc-AxhCb ffaSdK k-Carousel__page"
      >
        <div
          className="sc-AxhUy hZqsUu k-Carousel__page__item"
        >
          <div
            className="k-ProjectCard k-Card k-Card--light"
            href={null}
            title={null}
          >
            <div
              className="k-ProjectCard__grid"
            >
              <div
                className="sc-fzqNJr fSnHqq k-ProjectCard__grid--flex"
              >
                <span
                  className="k-ButtonImage k-ButtonImage--withoutPointerEvents"
                  tabIndex={null}
                >
                  <img
                    alt=""
                    className="k-ButtonImage__img"
                    src={null}
                  />
                </span>
                <p
                  className="k-Paragraph k-ProjectCard__grid--flex__item-fluid k-u-margin-left-single k-u-margin-right-single k-Paragraph--quaternary k-Paragraph--withoutMargin k-Paragraph--normalLineHeight"
                >
                  <span
                    className="k-u-weight-regular"
                  >
                    Name
                  </span>
                </p>
              </div>
              <div
                className="sc-fzqNJr jsUow k-ProjectCard__grid--flex"
              >
                <p
                  className="sc-fzoyAV hEKEUl k-ProjectCard__grid--flex__item-fluid k-Card__title"
                >
                  A
                </p>
              </div>
            </div>
          </div>
        </div>
        <div
          className="sc-AxhUy iLqwOC k-Carousel__page__item"
        />
        <div
          className="sc-AxhUy iLqwOC k-Carousel__page__item"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<Carousel /> with loop prop on desktop matches with snapshot 1`] = `
<div
  className="sc-fzozJi dfRDUK k-Carousel"
>
  <div
    className="sc-AxgMl jWGKu k-Carousel__inner"
    onScroll={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="sc-AxheI fsZQch k-Carousel__inner__pageContainer"
      onClick={[Function]}
    >
      <div
        className="sc-AxhCb ffaSdK k-Carousel__page"
      >
        <div
          className="sc-AxhUy hZqsUu k-Carousel__page__item"
        >
          <div
            className="k-ProjectCard k-Card k-Card--light"
            href={null}
            title={null}
          >
            <div
              className="k-ProjectCard__grid"
            >
              <div
                className="sc-fzqNJr fSnHqq k-ProjectCard__grid--flex"
              >
                <span
                  className="k-ButtonImage k-ButtonImage--withoutPointerEvents"
                  tabIndex={null}
                >
                  <img
                    alt=""
                    className="k-ButtonImage__img"
                    src={null}
                  />
                </span>
                <p
                  className="k-Paragraph k-ProjectCard__grid--flex__item-fluid k-u-margin-left-single k-u-margin-right-single k-Paragraph--quaternary k-Paragraph--withoutMargin k-Paragraph--normalLineHeight"
                >
                  <span
                    className="k-u-weight-regular"
                  >
                    Name
                  </span>
                </p>
              </div>
              <div
                className="sc-fzqNJr jsUow k-ProjectCard__grid--flex"
              >
                <p
                  className="sc-fzoyAV hEKEUl k-ProjectCard__grid--flex__item-fluid k-Card__title"
                >
                  A
                </p>
              </div>
            </div>
          </div>
        </div>
        <div
          className="sc-AxhUy ezsiGz k-Carousel__page__item"
        />
        <div
          className="sc-AxhUy ezsiGz k-Carousel__page__item"
        />
      </div>
    </div>
  </div>
</div>
`;
